### PR TITLE
Add ASB defaults

### PR DESCRIPTION
**Pull request by WilliamBZA**
Closes https://github.com/Particular/ServiceControl/issues/1308

Adds the defaults identified by @yvesgoeleven in #1308

### CODE REVIEW

File: src/ServiceControl.Transports.ASB/ASBEndpointTopologyTransportCustomization.cs [hunk 1]
```diff
@@ -37,12 +39,18 @@ public override void CustomizeRawEndpoint(RawEndpointConfiguration endpointConfi
...
+            endpointConfig.LimitMessageProcessingConcurrencyTo(Math.Min(Environment.ProcessorCount, transportSettings.MaxConcurrency));
         }
 
         static void ConfigureTransport(TransportExtensions<AzureServiceBusTransport> transport, TransportSettings transportSettings)
         {
             transport.Transactions(TransportTransactionMode.SendsAtomicWithReceive);
             transport.ConnectionString(transportSettings.ConnectionString);
+
+            transport.MessageReceivers().PrefetchCount(0);
+            transport.MessageReceivers().AutoRenewTimeout(TimeSpan.FromMinutes(5));
```
Comment (yvesgoeleven) [inline src/ServiceControl.Transports.ASB/ASBEndpointTopologyTransportCustomization.cs]: No, don't set autorenew, I meant Queues.LockDuration & Subscriptions.LockDuration

File: src/ServiceControl.Transports.ASB/ASBEndpointTopologyTransportCustomization.cs [hunk 2]
```diff
@@ -37,12 +39,18 @@ public override void CustomizeRawEndpoint(RawEndpointConfiguration endpointConfi
             transport.UseEndpointOrientedTopology();
             transport.ApplyHacksForNsbRaw();
             ConfigureTransport(transport, transportSettings);
+
+            endpointConfig.LimitMessageProcessingConcurrencyTo(Math.Min(Environment.ProcessorCount, transportSettings.MaxConcurrency));
```
Comment (yvesgoeleven) [inline src/ServiceControl.Transports.ASB/ASBEndpointTopologyTransportCustomization.cs]: This should be MessageReceivers().NumberOfClientsPerEntity(Math.Min(Environment.ProcessorCount, transportSettings.MaxConcurrency))

Comment (yvesgoeleven) [reply src/ServiceControl.Transports.ASB/ASBEndpointTopologyTransportCustomization.cs]: probably want to move it to ConfigureTransport as well

File: src/ServiceControl.Transports.ASB/ASBEndpointTopologyTransportCustomization.cs [hunk 3]
```diff
@@ -29,6 +29,8 @@ public override void CustomizeEndpoint(EndpointConfiguration endpointConfig, Tra
             }
 
             ConfigureTransport(transport, transportSettings);
+
+            endpointConfig.LimitMessageProcessingConcurrencyTo(Math.Min(Environment.ProcessorCount, transportSettings.MaxConcurrency));
```
Comment (yvesgoeleven) [inline src/ServiceControl.Transports.ASB/ASBEndpointTopologyTransportCustomization.cs]: Same as below

File: src/ServiceControl.Transports.ASB/ASBForwardingTopologyTransportCustomization.cs [hunk 4]
```diff
@@ -9,13 +10,17 @@ public override void CustomizeEndpoint(EndpointConfiguration endpointConfig, Tra
         {
             var transport = endpointConfig.UseTransport<AzureServiceBusTransport>();
             ConfigureTransport(transport, transportSettings);
+
+            endpointConfig.LimitMessageProcessingConcurrencyTo(Math.Min(Environment.ProcessorCount, transportSettings.MaxConcurrency));
```
Comment (yvesgoeleven) [inline src/ServiceControl.Transports.ASB/ASBForwardingTopologyTransportCustomization.cs]: Same as above

File: src/ServiceControl.Transports.ASB/ASBForwardingTopologyTransportCustomization.cs [hunk 5]
```diff
@@ -9,13 +10,17 @@ public override void CustomizeEndpoint(EndpointConfiguration endpointConfig, Tra
...
         }
 
         public override void CustomizeRawEndpoint(RawEndpointConfiguration endpointConfig, TransportSettings transportSettings)
         {
             var transport = endpointConfig.UseTransport<AzureServiceBusTransport>();
             transport.ApplyHacksForNsbRaw();
             ConfigureTransport(transport, transportSettings);
+
+            endpointConfig.LimitMessageProcessingConcurrencyTo(Math.Min(Environment.ProcessorCount, transportSettings.MaxConcurrency));
```
Comment (yvesgoeleven) [inline src/ServiceControl.Transports.ASB/ASBForwardingTopologyTransportCustomization.cs]: Same as above

File: src/ServiceControl.Transports.ASB/ASBForwardingTopologyTransportCustomization.cs [hunk 6]
```diff
@@ -24,6 +29,10 @@ static void ConfigureTransport(TransportExtensions<AzureServiceBusTransport> tra
             transport.Transactions(TransportTransactionMode.SendsAtomicWithReceive);
             transport.ConnectionString(transportSettings.ConnectionString);
             transport.Sanitization().UseStrategy<ValidateAndHashIfNeeded>();
+
+            transport.MessageReceivers().PrefetchCount(0);
+            transport.MessageReceivers().AutoRenewTimeout(TimeSpan.FromMinutes(5));
```
Comment (yvesgoeleven) [inline src/ServiceControl.Transports.ASB/ASBForwardingTopologyTransportCustomization.cs]: Same as above

**Comment by WilliamBZA** [comment]
Updated @yvesgoeleven

**Comment by yvesgoeleven** [comment]
Just a question as my understanding of SC is limited. settings.MaximumConcurrencyLevel is SC's concurrency level right?

**Comment by WilliamBZA** [comment]
Yes, which is also passed through to the NServiceBus endpointConfiguration's `LimitMessageProcessingConcurrencyTo` method

**Comment by yvesgoeleven** [comment]
perfect!

File: src/ServiceControl.Transports.ASB/ASBForwardingTopologyTransportCustomization.cs [hunk 7]
```diff
@@ -24,6 +25,12 @@ static void ConfigureTransport(TransportExtensions<AzureServiceBusTransport> tra
             transport.Transactions(TransportTransactionMode.SendsAtomicWithReceive);
             transport.ConnectionString(transportSettings.ConnectionString);
             transport.Sanitization().UseStrategy<ValidateAndHashIfNeeded>();
+
+            transport.MessageReceivers().PrefetchCount(0);
+            transport.Queues().LockDuration(TimeSpan.FromMinutes(5));
+            transport.Subscriptions().LockDuration(TimeSpan.FromMinutes(5));
+            transport.MessagingFactories().NumberOfMessagingFactoriesPerNamespace(2);
+            transport.NumberOfClientsPerEntity(Math.Min(Environment.ProcessorCount, transportSettings.MaxConcurrency));
```
Comment (mikeminutillo) [inline src/ServiceControl.Transports.ASB/ASBForwardingTopologyTransportCustomization.cs]: Should we move these into a base class or a shared location? Or is the fact that they are currently identical a coincidence?

Comment (SzymonPobiega) [reply src/ServiceControl.Transports.ASB/ASBForwardingTopologyTransportCustomization.cs]: I guess these are not topology-dependent so 👍 for a base class

**Pull request merged**